autosklearn/automl.py

```diff
@@ -1834,15 +1834,52 @@ def get_models_with_weights(self):
         return self.ensemble_.get_models_with_weights(self.models_)
 
     def show_models(self):
-        models_with_weights = self.get_models_with_weights()
-
-        with io.StringIO() as sio:
-            sio.write("[")
-            for weight, model in models_with_weights:
-                sio.write("(%f, %s),\n" % (weight, model))
-            sio.write("]")
-
-            return sio.getvalue()
+        """ Returns a dictionary containing models and their information
+            where model_id/run_id is the key """
+
+        ensemble_dict={} 
+        
+        def has_key(rv, key):
+            return rv.additional_info and key in rv.additional_info
+        
+        table_dict={}
+        for rkey, rval in self.runhistory_.data.items():
+            if has_key(rval, 'num_run'):
+                table_dict[rval.additional_info['num_run']]={
+                        'model_id': rval.additional_info['num_run'],
+                        'cost': rval.cost}
+                
+        for i, weight in enumerate(self.ensemble_.weights_):
+            (_, model_id, _) = self.ensemble_.identifiers_[i]
+            table_dict[model_id]['ensemble_weight']=weight
+        
+        table=pd.DataFrame.from_dict(table_dict, orient='index')
+        table.sort_values(by='cost', inplace=True)
+        table['rank']=range(1,len(table)+1)
+
+        for (_, model_id, _), model in self.models_.items():
```

Comment:
`self.models_` may be empty if a user specifies using cross validation, in which case `self.cv_models_` is set. We want to unify them into one thing but at the moment that doesn't happen.
```python
self.cv_models_
```

For this a test probably needs to be created, I will elaborate on creating a test in a moment.

Reply:
Did not know that. So do you want me to put some condition to check for this and use either `self.models_` or `self.cv_models_`  inside the loop depending on whichever is _non-empty_, or should I just create the test for now?

Reply:
Yeah, you can either check if the `resampling_strategy` is `cv` or more useful is probably just check which one is empty.